// carbon-components-angular/esm2020/tiles/tile.component.mjs
import { Component, HostBinding, Input } from "@angular/core";
import * as i0 from "@angular/core";
/**
 * Build application's tiles using this component. Get started with importing the module:
 *
 * ```typescript
 * import { TilesModule } from 'carbon-components-angular';
 * ```
 *
 * ```html
 * <cds-tile>
 * 		tile content
 * </cds-tile>
 * ```
 *
 * [See demo](../../?path=/story/components-tiles--basic)
 */
export class Tile {
    constructor() {
        this.tileClass = true;
        /**
         * @deprecated since v5 - Use `cdsLayer` directive instead
         * Set to `"light"` to apply the light style
         */
        this.theme = "dark";
    }
    get lightThemeEnabled() {
        return this.theme === "light";
    }
}
Tile.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: Tile, deps: [], target: i0.ɵɵFactoryTarget.Component });
Tile.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: Tile, selector: "cds-tile, ibm-tile", inputs: { theme: "theme" }, host: { properties: { "class.cds--tile": "this.tileClass", "class.cds--tile--light": "this.lightThemeEnabled" } }, ngImport: i0, template: `<ng-content></ng-content>`, isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: Tile, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-tile, ibm-tile",
                    template: `<ng-content></ng-content>`
                }]
        }], propDecorators: { tileClass: [{
                type: HostBinding,
                args: ["class.cds--tile"]
            }], lightThemeEnabled: [{
                type: HostBinding,
                args: ["class.cds--tile--light"]
            }], theme: [{
                type: Input
            }] } });
