// carbon-components-angular/esm2020/table/head/table-head-expand.component.mjs
import { Component, EventEmitter, HostBinding, Input, Output } from "@angular/core";
import * as i0 from "@angular/core";
import * as i1 from "carbon-components-angular/i18n";
import * as i2 from "@angular/common";
import * as i3 from "carbon-components-angular/icon";
export class TableHeadExpand {
    constructor(i18n) {
        this.i18n = i18n;
        this.hostClass = true;
        this.showExpandAllToggle = false;
        this.expanded = false;
        this.expandedChange = new EventEmitter();
        this._ariaLabel = this.i18n.getOverridable("TABLE.EXPAND_ALL_BUTTON");
    }
    get previousValue() {
        return this.expanded ? "collapsed" : null;
    }
    getAriaLabel() {
        return this._ariaLabel.subject;
    }
}
TableHeadExpand.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableHeadExpand, deps: [{ token: i1.I18n }], target: i0.ɵɵFactoryTarget.Component });
TableHeadExpand.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: TableHeadExpand, selector: "[cdsTableHeadExpand], [ibmTableHeadExpand]", inputs: { showExpandAllToggle: "showExpandAllToggle", expanded: "expanded" }, outputs: { expandedChange: "expandedChange" }, host: { properties: { "class.cds--table-expand": "this.hostClass", "attr.data-previous-value": "this.previousValue" } }, ngImport: i0, template: `
		<button
			*ngIf="showExpandAllToggle"
			class="cds--table-expand__button"
			[attr.aria-label]="getAriaLabel() | async"
			(click)="expandedChange.emit(!expanded)">
			<svg cdsIcon="chevron--right" size="16" class="cds--table-expand__svg"></svg>
		</button>
		<ng-container *ngIf="!showExpandAllToggle">
			<ng-content></ng-content>
		</ng-container>
	`, isInline: true, dependencies: [{ kind: "directive", type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i3.IconDirective, selector: "[cdsIcon], [ibmIcon]", inputs: ["ibmIcon", "cdsIcon", "size", "title", "ariaLabel", "ariaLabelledBy", "ariaHidden", "isFocusable"] }, { kind: "pipe", type: i2.AsyncPipe, name: "async" }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TableHeadExpand, decorators: [{
            type: Component,
            args: [{
                    // tslint:disable-next-line: component-selector
                    selector: "[cdsTableHeadExpand], [ibmTableHeadExpand]",
                    template: `
		<button
			*ngIf="showExpandAllToggle"
			class="cds--table-expand__button"
			[attr.aria-label]="getAriaLabel() | async"
			(click)="expandedChange.emit(!expanded)">
			<svg cdsIcon="chevron--right" size="16" class="cds--table-expand__svg"></svg>
		</button>
		<ng-container *ngIf="!showExpandAllToggle">
			<ng-content></ng-content>
		</ng-container>
	`
                }]
        }], ctorParameters: function () { return [{ type: i1.I18n }]; }, propDecorators: { hostClass: [{
                type: HostBinding,
                args: ["class.cds--table-expand"]
            }], showExpandAllToggle: [{
                type: Input
            }], expanded: [{
                type: Input
            }], expandedChange: [{
                type: Output
            }], previousValue: [{
                type: HostBinding,
                args: ["attr.data-previous-value"]
            }] } });
