// carbon-components-angular/esm2020/tag/tag-selectable.component.mjs
import { ChangeDetectionStrategy, Component, EventEmitter, HostBinding, HostListener, Input, Output } from "@angular/core";
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class TagSelectableComponent {
    constructor() {
        this.role = "button";
        this.buttonType = "button";
        this.tabIndex = 0;
        this.size = "md";
        this.skeleton = false;
        this.disabled = false;
        this.class = "";
        this.selected = false;
        this.selectedChange = new EventEmitter();
    }
    get ariaPressed() {
        return this.selected;
    }
    onClick() {
        this.selected = !this.selected;
        this.selectedChange.emit(this.selected);
    }
    /**
     * @todo
     * Remove `cds--tag--${this.size}` in v7
     */
    get attrClass() {
        const disabledClass = this.disabled ? "cds--tag--disabled" : "";
        const sizeClass = `cds--tag--${this.size} cds--layout--size-${this.size}`;
        const skeletonClass = this.skeleton ? "cds--skeleton" : "";
        const selectedClass = this.selected ? "cds--tag--selectable-selected" : "";
        return `cds--tag cds--tag--selectable ${selectedClass} ${disabledClass} ${sizeClass} ${skeletonClass} ${this.class}`;
    }
}
TagSelectableComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TagSelectableComponent, deps: [], target: i0.ɵɵFactoryTarget.Component });
TagSelectableComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: TagSelectableComponent, selector: "cds-tag-selectable, ibm-tag-selectable", inputs: { size: "size", skeleton: "skeleton", disabled: "disabled", class: "class", selected: "selected" }, outputs: { selectedChange: "selectedChange" }, host: { listeners: { "click": "onClick()" }, properties: { "attr.role": "this.role", "attr.type": "this.buttonType", "attr.tabindex": "this.tabIndex", "attr.aria-pressed": "this.ariaPressed", "attr.class": "this.attrClass" } }, ngImport: i0, template: `
		<ng-container *ngIf="!skeleton">
			<ng-content select="[cdsTagIcon],[ibmTagIcon]"></ng-content>
			<span class="cds--tag__label">
				<ng-content></ng-content>
			</span>
		</ng-container>
	`, isInline: true, dependencies: [{ kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TagSelectableComponent, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-tag-selectable, ibm-tag-selectable",
                    template: `
		<ng-container *ngIf="!skeleton">
			<ng-content select="[cdsTagIcon],[ibmTagIcon]"></ng-content>
			<span class="cds--tag__label">
				<ng-content></ng-content>
			</span>
		</ng-container>
	`,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], propDecorators: { role: [{
                type: HostBinding,
                args: ["attr.role"]
            }], buttonType: [{
                type: HostBinding,
                args: ["attr.type"]
            }], tabIndex: [{
                type: HostBinding,
                args: ["attr.tabindex"]
            }], ariaPressed: [{
                type: HostBinding,
                args: ["attr.aria-pressed"]
            }], size: [{
                type: Input
            }], skeleton: [{
                type: Input
            }], disabled: [{
                type: Input
            }], class: [{
                type: Input
            }], selected: [{
                type: Input
            }], selectedChange: [{
                type: Output
            }], onClick: [{
                type: HostListener,
                args: ["click"]
            }], attrClass: [{
                type: HostBinding,
                args: ["attr.class"]
            }] } });
