// carbon-components-angular/esm2020/tag/tag-operational.component.mjs
import { ChangeDetectionStrategy, Component, HostBinding, Input } from "@angular/core";
import { Tag } from "./tag.component";
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
export class TagOperationalComponent extends Tag {
    constructor() {
        super(...arguments);
        this.role = "button";
        this.buttonType = "button";
        this.tabIndex = 0;
        this.disabled = false;
    }
    /**
     * @todo
     * Remove `cds--tag--${this.size}` in v7
     */
    get attrClass() {
        const disabledClass = this.disabled ? "cds--tag--disabled" : "";
        const sizeClass = `cds--tag--${this.size} cds--layout--size-${this.size}`;
        const skeletonClass = this.skeleton ? "cds--skeleton" : "";
        return `cds--tag cds--tag--operational cds--tag--${this.type} ${disabledClass} ${sizeClass} ${skeletonClass} ${this.class}`;
    }
}
TagOperationalComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TagOperationalComponent, deps: null, target: i0.ɵɵFactoryTarget.Component });
TagOperationalComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: TagOperationalComponent, selector: "cds-tag-operational, ibm-tag-operational", inputs: { disabled: "disabled" }, host: { properties: { "attr.role": "this.role", "attr.type": "this.buttonType", "attr.tabindex": "this.tabIndex", "attr.class": "this.attrClass" } }, usesInheritance: true, ngImport: i0, template: `
		<ng-container *ngIf="!skeleton">
			<ng-content select="[cdsTagIcon],[ibmTagIcon]"></ng-content>
			<span class="cds--tag__label">
				<ng-content></ng-content>
			</span>
		</ng-container>
	`, isInline: true, dependencies: [{ kind: "directive", type: i1.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TagOperationalComponent, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-tag-operational, ibm-tag-operational",
                    template: `
		<ng-container *ngIf="!skeleton">
			<ng-content select="[cdsTagIcon],[ibmTagIcon]"></ng-content>
			<span class="cds--tag__label">
				<ng-content></ng-content>
			</span>
		</ng-container>
	`,
                    changeDetection: ChangeDetectionStrategy.OnPush
                }]
        }], propDecorators: { role: [{
                type: HostBinding,
                args: ["attr.role"]
            }], buttonType: [{
                type: HostBinding,
                args: ["attr.type"]
            }], tabIndex: [{
                type: HostBinding,
                args: ["attr.tabindex"]
            }], disabled: [{
                type: Input
            }], attrClass: [{
                type: HostBinding,
                args: ["attr.class"]
            }] } });
