// carbon-components-angular/esm2020/tag/tag.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { Tag } from "./tag.component";
import { TagFilter } from "./tag-filter.component";
import { IconModule } from "carbon-components-angular/icon";
import { TagIconDirective } from "./tag-icon.directive";
import { TagSelectableComponent } from "./tag-selectable.component";
import { TagOperationalComponent } from "./tag-operational.component";
import * as i0 from "@angular/core";
export class TagModule {
}
TagModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TagModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TagModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: TagModule, declarations: [Tag,
        TagFilter,
        TagIconDirective,
        TagSelectableComponent,
        TagOperationalComponent], imports: [CommonModule, IconModule], exports: [Tag,
        TagFilter,
        TagIconDirective,
        TagSelectableComponent,
        TagOperationalComponent] });
TagModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TagModule, imports: [CommonModule, IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TagModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Tag,
                        TagFilter,
                        TagIconDirective,
                        TagSelectableComponent,
                        TagOperationalComponent
                    ],
                    exports: [
                        Tag,
                        TagFilter,
                        TagIconDirective,
                        TagSelectableComponent,
                        TagOperationalComponent
                    ],
                    imports: [CommonModule, IconModule]
                }]
        }] });
