// carbon-components-angular/esm2020/button/button-set.component.mjs
import { Component, HostBinding } from "@angular/core";
import * as i0 from "@angular/core";
/**
 * Get started with importing the module:
 *
 * ```typescript
 * import { ButtonModule } from 'carbon-components-angular';
 * ```
 *
 * [See demo](../../?path=/story/components-button-button-set--basic)
 */
export class ButtonSet {
    constructor() {
        this.buttonSetClass = true;
    }
}
ButtonSet.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ButtonSet, deps: [], target: i0.ɵɵFactoryTarget.Component });
ButtonSet.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: ButtonSet, selector: "cds-button-set, ibm-button-set", host: { properties: { "class.cds--btn-set": "this.buttonSetClass" } }, ngImport: i0, template: "<ng-content></ng-content>", isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: ButtonSet, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-button-set, ibm-button-set",
                    template: "<ng-content></ng-content>"
                }]
        }], propDecorators: { buttonSetClass: [{
                type: HostBinding,
                args: ["class.cds--btn-set"]
            }] } });
