// carbon-components-angular/esm2020/link/link.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { Link } from "./link.directive";
import { LinkIconDirective } from "./link-icon.directive";
import * as i0 from "@angular/core";
export class LinkModule {
}
LinkModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LinkModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LinkModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: LinkModule, declarations: [Link,
        LinkIconDirective], imports: [CommonModule], exports: [Link,
        LinkIconDirective] });
LinkModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LinkModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: LinkModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Link,
                        LinkIconDirective
                    ],
                    exports: [
                        Link,
                        LinkIconDirective
                    ],
                    imports: [
                        CommonModule
                    ]
                }]
        }] });
