// carbon-components-angular/esm2020/ui-shell/sidenav/sidenav-item.component.mjs
import { Component, Input, Optional, Output, EventEmitter, HostBinding } from "@angular/core";
import * as i0 from "@angular/core";
import * as i1 from "@angular/platform-browser";
import * as i2 from "@angular/router";
import * as i3 from "@angular/common";
import * as i4 from "./routerlink-extended.directive";
/**
 * `SideNavItem` can either be a child of `SideNav` or `SideNavMenu`
 */
export class SideNavItem {
    constructor(domSanitizer, router) {
        this.domSanitizer = domSanitizer;
        this.router = router;
        /**
         * Use the routerLink attribute on <a> tag for navigation instead of using event handlers
         */
        this.useRouter = false;
        /**
         * Toggles the active (current page) state for the link.
         */
        this.active = false;
        this.isSubMenu = false;
        /**
         * Emits the navigation status promise when the link is activated
         */
        this.navigation = new EventEmitter();
        /**
         * Emits when `active` input is changed. This is mainly used to indicate to any parent menu items that a
         * child sidenav item is active or not active.
         */
        this.selected = new EventEmitter();
        this.role = "listitem";
        this._href = "#";
    }
    /**
     * Link for the item. NOTE: *do not* pass unsafe or untrusted values, this has the potential to open you up to XSS attacks
     */
    set href(v) {
        // Needed when component is created dynamically with a model.
        if (v === undefined) {
            return;
        }
        this._href = v;
    }
    get href() {
        return this.domSanitizer.bypassSecurityTrustUrl(this._href);
    }
    get sideNav() {
        return !this.isSubMenu;
    }
    get menuItem() {
        return this.isSubMenu;
    }
    ngOnChanges(changes) {
        if (changes.active) {
            this.selected.emit(this.active);
        }
    }
    navigate(event) {
        if (this.router && this.route) {
            event.preventDefault();
            const status = this.router.navigate(this.route, this.routeExtras);
            this.navigation.emit(status);
        }
        else if (this._href === "#") {
            event.preventDefault();
        }
    }
}
SideNavItem.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SideNavItem, deps: [{ token: i1.DomSanitizer }, { token: i2.Router, optional: true }], target: i0.ɵɵFactoryTarget.Component });
SideNavItem.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: SideNavItem, selector: "cds-sidenav-item, ibm-sidenav-item", inputs: { href: "href", useRouter: "useRouter", active: "active", route: "route", isSubMenu: "isSubMenu", routeExtras: "routeExtras", title: "title" }, outputs: { navigation: "navigation", selected: "selected" }, host: { properties: { "class.cds--side-nav__item": "this.sideNav", "class.cds--side-nav__menu-item": "this.menuItem", "attr.role": "this.role" } }, usesOnChanges: true, ngImport: i0, template: `
		<a *ngIf="!useRouter; else sidenavItemRouterTpl"
			class="cds--side-nav__link"
			[ngClass]="{
				'cds--side-nav__item--active': active
			}"
			[href]="href"
			[attr.aria-current]="(active ? 'page' : null)"
			[attr.title]="title ? title : null"
			(click)="navigate($event)">
			<ng-template [ngTemplateOutlet]="sidenavItemContentTpl"></ng-template>
		</a>

		<ng-template #sidenavItemRouterTpl>
			<a
				[attr.title]="title ? title : null"
				[routerLink]="route"
				[routeExtras]="routeExtras"
				routerLinkActive="cds--side-nav__item--active"
				ariaCurrentWhenActive="page"
				class="cds--side-nav__link">
				<ng-template [ngTemplateOutlet]="sidenavItemContentTpl"></ng-template>
			</a>
		</ng-template>

		<ng-template #sidenavItemContentTpl>
			<div *ngIf="!isSubMenu" class="cds--side-nav__icon">
				<ng-content select="svg, [icon]"></ng-content>
			</div>
			<span class="cds--side-nav__link-text">
				<ng-content></ng-content>
			</span>
		</ng-template>
	`, isInline: true, styles: [":host{display:list-item}\n"], dependencies: [{ kind: "directive", type: i3.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i3.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }, { kind: "directive", type: i2.RouterLinkWithHref, selector: "a[routerLink],area[routerLink]", inputs: ["target", "queryParams", "fragment", "queryParamsHandling", "state", "relativeTo", "preserveFragment", "skipLocationChange", "replaceUrl", "routerLink"] }, { kind: "directive", type: i2.RouterLinkActive, selector: "[routerLinkActive]", inputs: ["routerLinkActiveOptions", "ariaCurrentWhenActive", "routerLinkActive"], outputs: ["isActiveChange"], exportAs: ["routerLinkActive"] }, { kind: "directive", type: i4.RouterLinkExtendedDirective, selector: "[routerLink]", inputs: ["routeExtras"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SideNavItem, decorators: [{
            type: Component,
            args: [{ selector: "cds-sidenav-item, ibm-sidenav-item", template: `
		<a *ngIf="!useRouter; else sidenavItemRouterTpl"
			class="cds--side-nav__link"
			[ngClass]="{
				'cds--side-nav__item--active': active
			}"
			[href]="href"
			[attr.aria-current]="(active ? 'page' : null)"
			[attr.title]="title ? title : null"
			(click)="navigate($event)">
			<ng-template [ngTemplateOutlet]="sidenavItemContentTpl"></ng-template>
		</a>

		<ng-template #sidenavItemRouterTpl>
			<a
				[attr.title]="title ? title : null"
				[routerLink]="route"
				[routeExtras]="routeExtras"
				routerLinkActive="cds--side-nav__item--active"
				ariaCurrentWhenActive="page"
				class="cds--side-nav__link">
				<ng-template [ngTemplateOutlet]="sidenavItemContentTpl"></ng-template>
			</a>
		</ng-template>

		<ng-template #sidenavItemContentTpl>
			<div *ngIf="!isSubMenu" class="cds--side-nav__icon">
				<ng-content select="svg, [icon]"></ng-content>
			</div>
			<span class="cds--side-nav__link-text">
				<ng-content></ng-content>
			</span>
		</ng-template>
	`, styles: [":host{display:list-item}\n"] }]
        }], ctorParameters: function () { return [{ type: i1.DomSanitizer }, { type: i2.Router, decorators: [{
                    type: Optional
                }] }]; }, propDecorators: { href: [{
                type: Input
            }], useRouter: [{
                type: Input
            }], sideNav: [{
                type: HostBinding,
                args: ["class.cds--side-nav__item"]
            }], menuItem: [{
                type: HostBinding,
                args: ["class.cds--side-nav__menu-item"]
            }], active: [{
                type: Input
            }], route: [{
                type: Input
            }], isSubMenu: [{
                type: Input
            }], routeExtras: [{
                type: Input
            }], title: [{
                type: Input
            }], navigation: [{
                type: Output
            }], selected: [{
                type: Output
            }], role: [{
                type: HostBinding,
                args: ["attr.role"]
            }] } });
