// carbon-components-angular/esm2020/ui-shell/sidenav/sidenav.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { RouterModule } from "@angular/router";
import { I18nModule } from "carbon-components-angular/i18n";
import { SideNav } from "./sidenav.component";
import { SideNavItem } from "./sidenav-item.component";
import { SideNavMenu } from "./sidenav-menu.component";
import { RouterLinkExtendedDirective } from "./routerlink-extended.directive";
import * as i0 from "@angular/core";
export { SideNav, SideNavItem, SideNavMenu };
export class SideNavModule {
}
SideNavModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SideNavModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
SideNavModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: SideNavModule, declarations: [SideNav,
        SideNavItem,
        SideNavMenu,
        RouterLinkExtendedDirective], imports: [CommonModule, I18nModule, RouterModule], exports: [SideNav,
        SideNavItem,
        SideNavMenu] });
SideNavModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SideNavModule, imports: [CommonModule, I18nModule, RouterModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: SideNavModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        SideNav,
                        SideNavItem,
                        SideNavMenu,
                        RouterLinkExtendedDirective
                    ],
                    imports: [CommonModule, I18nModule, RouterModule],
                    exports: [
                        SideNav,
                        SideNavItem,
                        SideNavMenu
                    ]
                }]
        }] });
