// carbon-components-angular/esm2020/treeview/treeview.module.mjs
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { IconModule } from "carbon-components-angular/icon";
import { TreeViewComponent } from "./treeview.component";
import { TreeNodeComponent } from "./tree-node.component";
import * as i0 from "@angular/core";
export class TreeviewModule {
}
TreeviewModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TreeviewModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
TreeviewModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: TreeviewModule, declarations: [TreeViewComponent, TreeNodeComponent], imports: [CommonModule, IconModule], exports: [TreeViewComponent, TreeNodeComponent] });
TreeviewModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TreeviewModule, imports: [CommonModule, IconModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TreeviewModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [TreeViewComponent, TreeNodeComponent],
                    exports: [TreeViewComponent, TreeNodeComponent],
                    imports: [CommonModule, IconModule]
                }]
        }] });
