// carbon-components-angular/esm2020/treeview/treeview.component.mjs
import { DOCUMENT } from "@angular/common";
import { Component, Input, Output, TemplateRef, EventEmitter, Inject, ViewChild } from "@angular/core";
import { TreeViewService } from "./treeview.service";
import * as i0 from "@angular/core";
import * as i1 from "./treeview.service";
import * as i2 from "@angular/common";
import * as i3 from "./tree-node.component";
/**
 * Get started with importing the module:
 *
 * ```typescript
 * import { TreeviewModule } from 'carbon-components-angular';
 * ```
 *
 * [See demo](../../?path=/story/components-tree-view--basic)
 */
export class TreeViewComponent {
    constructor(document, treeViewService, elementRef) {
        this.document = document;
        this.treeViewService = treeViewService;
        this.elementRef = elementRef;
        this.id = `tree-view-${TreeViewComponent.treeViewCount++}`;
        /**
         * Specify the size of the list items in the tree
         */
        this.size = "sm";
        this.select = new EventEmitter();
        this.toggle = new EventEmitter();
        this._tree = [];
    }
    /**
     * Pass `Node[]` array to have tree view render the nodes
     * Passing value will disregard projected content
     */
    set tree(treeNodes) {
        this._tree = treeNodes.map((node) => this.copyNode(node));
        this.treeViewService.contentProjected = false;
    }
    get tree() {
        return this._tree;
    }
    /**
     * **Experimental** - Enable to select multiple nodes
     */
    set isMultiSelect(isMulti) {
        this.treeViewService.isMultiSelect = isMulti;
    }
    /**
     * Subscribe for node selection
     */
    ngOnInit() {
        this.subscription = this.treeViewService.selectionObservable.subscribe((nodesMap) => {
            // Get all values from the map to emit
            const nodes = [...nodesMap.values()];
            this.select.emit(this.treeViewService.isMultiSelect ? nodes : nodes[0]);
        });
        this.subscription.add(this.treeViewService.focusNodeObservable.subscribe(node => this.onNodeFocusChange(node)));
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
    /**
     * Initialize tree walker to support keyboard navigation
     */
    ngAfterViewInit() {
        this.treeWalker = this.document.createTreeWalker(this.root.nativeElement, NodeFilter.SHOW_ELEMENT, {
            acceptNode: function (node) {
                if (node.classList.contains(`cds--tree-node--disabled`)) {
                    return NodeFilter.FILTER_REJECT;
                }
                if (node.matches(`div.cds--tree-node`)) {
                    return NodeFilter.FILTER_ACCEPT;
                }
                return NodeFilter.FILTER_SKIP;
            }
        });
    }
    /**
     * Navigate tree using tree walker
     * @param event - KeyboardEvent
     */
    navigateTree(event) {
        if (event.key === "ArrowUp") {
            this.treeWalker.previousNode()?.focus();
        }
        if (event.key === "ArrowDown") {
            this.treeWalker.nextNode()?.focus();
        }
    }
    /**
     * Propagate node toggle event
     * @param eventOnNode - EventOnNode
     */
    onNodeToggle(eventOnNode) {
        if (!eventOnNode) {
            return;
        }
        this.toggle.emit(eventOnNode.node);
    }
    /**
     * Node focus change
     * @param node - Node
     */
    onNodeFocusChange(node) {
        if (!node) {
            // if for some reason the focused node is not defined we fallback on the root element of the treeview
            this.treeWalker.currentNode = this.treeWalker.root;
            return;
        }
        // Update current node based on focus change to have a better keyboard navigation experience
        this.treeWalker.currentNode = this.elementRef.nativeElement.querySelector(`#${CSS.escape(node.id)}`);
    }
    isTemplate(value) {
        return value instanceof TemplateRef;
    }
    isProjected() {
        return this.treeViewService.contentProjected;
    }
    copyNode(node) {
        // making a recursive shallow copy to avoid performance issues when deeply cloning templateRefs if defined in the node
        const copiedNode = Object.assign({}, node);
        if (node.children) {
            copiedNode.children = node.children.map(child => this.copyNode(child));
        }
        return copiedNode;
    }
}
TreeViewComponent.treeViewCount = 0;
TreeViewComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TreeViewComponent, deps: [{ token: DOCUMENT }, { token: i1.TreeViewService }, { token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Component });
TreeViewComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: TreeViewComponent, selector: "cds-tree-view", inputs: { tree: "tree", id: "id", label: "label", labelContext: "labelContext", size: "size", isMultiSelect: "isMultiSelect" }, outputs: { select: "select", toggle: "toggle" }, providers: [TreeViewService], viewQueries: [{ propertyName: "root", first: true, predicate: ["treeWrapper"], descendants: true }], ngImport: i0, template: `
		<label
			*ngIf="label"
			[id]="id"
			class="cds--label">
			<ng-container *ngIf="!isTemplate(label)">{{label}}</ng-container>
			<ng-template
				*ngIf="isTemplate(label)"
				[ngTemplateOutlet]="label"
				[ngTemplateOutletContext]="{ $implicit: labelContext }">
			</ng-template>
		</label>
		<div
			class="cds--tree"
			[ngClass]="{
				'cds--tree--sm': size === 'sm',
				'cds--tree--xs': size === 'xs'
			}"
			[attr.aria-label]="label ? label : null"
			[attr.aria-labelledby]="!label ? id : null"
			[attr.aria-multiselectable]="isMultiSelect || null"
			role="tree"
			(keydown)="navigateTree($event)"
			#treeWrapper>
			<ng-container *ngIf="isProjected(); else notProjected">
				<ng-content></ng-content>
			</ng-container>
			<ng-template #notProjected>
				<cds-tree-node
					*ngFor="let node of tree"
					[node]="node"
					(nodetoggle)="onNodeToggle($event)">
				</cds-tree-node>
			</ng-template>
		</div>
	`, isInline: true, dependencies: [{ kind: "directive", type: i2.NgClass, selector: "[ngClass]", inputs: ["class", "ngClass"] }, { kind: "directive", type: i2.NgForOf, selector: "[ngFor][ngForOf]", inputs: ["ngForOf", "ngForTrackBy", "ngForTemplate"] }, { kind: "directive", type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { kind: "directive", type: i2.NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }, { kind: "component", type: i3.TreeNodeComponent, selector: "cds-tree-node", inputs: ["id", "active", "disabled", "selectable", "expanded", "label", "labelContext", "selected", "value", "icon", "iconContext", "gap", "children", "depth", "node"], outputs: ["nodeFocus", "nodeBlur", "nodeSelect", "nodetoggle"] }] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TreeViewComponent, decorators: [{
            type: Component,
            args: [{
                    selector: "cds-tree-view",
                    template: `
		<label
			*ngIf="label"
			[id]="id"
			class="cds--label">
			<ng-container *ngIf="!isTemplate(label)">{{label}}</ng-container>
			<ng-template
				*ngIf="isTemplate(label)"
				[ngTemplateOutlet]="label"
				[ngTemplateOutletContext]="{ $implicit: labelContext }">
			</ng-template>
		</label>
		<div
			class="cds--tree"
			[ngClass]="{
				'cds--tree--sm': size === 'sm',
				'cds--tree--xs': size === 'xs'
			}"
			[attr.aria-label]="label ? label : null"
			[attr.aria-labelledby]="!label ? id : null"
			[attr.aria-multiselectable]="isMultiSelect || null"
			role="tree"
			(keydown)="navigateTree($event)"
			#treeWrapper>
			<ng-container *ngIf="isProjected(); else notProjected">
				<ng-content></ng-content>
			</ng-container>
			<ng-template #notProjected>
				<cds-tree-node
					*ngFor="let node of tree"
					[node]="node"
					(nodetoggle)="onNodeToggle($event)">
				</cds-tree-node>
			</ng-template>
		</div>
	`,
                    providers: [TreeViewService]
                }]
        }], ctorParameters: function () { return [{ type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }, { type: i1.TreeViewService }, { type: i0.ElementRef }]; }, propDecorators: { tree: [{
                type: Input
            }], id: [{
                type: Input
            }], label: [{
                type: Input
            }], labelContext: [{
                type: Input
            }], size: [{
                type: Input
            }], isMultiSelect: [{
                type: Input
            }], select: [{
                type: Output
            }], toggle: [{
                type: Output
            }], root: [{
                type: ViewChild,
                args: ["treeWrapper"]
            }] } });
