// carbon-components-angular/esm2020/input/input.module.mjs
// modules
import { NgModule } from "@angular/core";
import { FormsModule } from "@angular/forms";
import { CommonModule } from "@angular/common";
// imports
import { Label } from "./label.component";
import { TextInput } from "./input.directive";
import { TextArea } from "./text-area.directive";
import { TextareaLabelComponent } from "./textarea-label.component";
import { TextInputLabelComponent } from "./text-input-label.component";
import { IconModule } from "carbon-components-angular/icon";
import { PasswordInput } from "./password.directive";
import { PasswordInputLabelComponent } from "./password-input-label.component";
import { TooltipModule } from "carbon-components-angular/tooltip";
import { ButtonModule } from "carbon-components-angular/button";
import * as i0 from "@angular/core";
export class InputModule {
}
InputModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: InputModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
InputModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: InputModule, declarations: [Label,
        TextInput,
        TextArea,
        PasswordInput,
        TextareaLabelComponent,
        TextInputLabelComponent,
        PasswordInputLabelComponent], imports: [CommonModule,
        FormsModule,
        IconModule,
        ButtonModule,
        TooltipModule], exports: [Label,
        TextareaLabelComponent,
        TextInputLabelComponent,
        PasswordInputLabelComponent,
        TextInput,
        TextArea,
        PasswordInput] });
InputModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: InputModule, imports: [CommonModule,
        FormsModule,
        IconModule,
        ButtonModule,
        TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: InputModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Label,
                        TextInput,
                        TextArea,
                        PasswordInput,
                        TextareaLabelComponent,
                        TextInputLabelComponent,
                        PasswordInputLabelComponent
                    ],
                    exports: [
                        Label,
                        TextareaLabelComponent,
                        TextInputLabelComponent,
                        PasswordInputLabelComponent,
                        TextInput,
                        TextArea,
                        PasswordInput
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        IconModule,
                        ButtonModule,
                        TooltipModule
                    ]
                }]
        }] });
