// carbon-components-angular/esm2020/input/input.directive.mjs
import { Directive, HostBinding, Input } from "@angular/core";
import * as i0 from "@angular/core";
/**
 * A directive for applying styling to an input element.
 *
 * Example:
 *
 * ```html
 * <input cdsText/>
 * ```
 *
 * See the [vanilla carbon docs](http://www.carbondesignsystem.com/components/text-input/code) for more detail.
 */
export class TextInput {
    constructor() {
        /**
         * @deprecated since v5 - Use `cdsLayer` directive instead
         * `light` or `dark` input theme
         */
        this.theme = "dark";
        /**
         * Input field render size
         */
        this.size = "md";
        this.inputClass = true;
        this.invalid = false;
        this.warn = false;
        this.skeleton = false;
    }
    /**
     * @todo - remove `cds--text-input--${size}` classes in v12
     */
    get isSizeSm() {
        return this.size === "sm";
    }
    get isSizeMd() {
        return this.size === "md";
    }
    get isSizelg() {
        return this.size === "lg";
    }
    // Size
    get sizeSm() {
        return this.size === "sm";
    }
    get sizeMd() {
        return this.size === "md";
    }
    get sizelg() {
        return this.size === "lg";
    }
    get isLightTheme() {
        return this.theme === "light";
    }
    get getInvalidAttribute() {
        return this.invalid ? true : undefined;
    }
}
TextInput.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TextInput, deps: [], target: i0.ɵɵFactoryTarget.Directive });
TextInput.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "14.3.0", type: TextInput, selector: "[cdsText], [ibmText]", inputs: { theme: "theme", size: "size", invalid: "invalid", warn: "warn", skeleton: "skeleton" }, host: { properties: { "class.cds--text-input": "this.inputClass", "class.cds--text-input--sm": "this.isSizeSm", "class.cds--text-input--md": "this.isSizeMd", "class.cds--text-input--lg": "this.isSizelg", "class.cds--layout--size-sm": "this.sizeSm", "class.cds--layout--size-md": "this.sizeMd", "class.cds--layout--size-lg": "this.sizelg", "class.cds--text-input--invalid": "this.invalid", "class.cds--text-input--warning": "this.warn", "class.cds--skeleton": "this.skeleton", "class.cds--text-input--light": "this.isLightTheme", "attr.data-invalid": "this.getInvalidAttribute" } }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: TextInput, decorators: [{
            type: Directive,
            args: [{
                    selector: "[cdsText], [ibmText]"
                }]
        }], propDecorators: { theme: [{
                type: Input
            }], size: [{
                type: Input
            }], inputClass: [{
                type: HostBinding,
                args: ["class.cds--text-input"]
            }], isSizeSm: [{
                type: HostBinding,
                args: ["class.cds--text-input--sm"]
            }], isSizeMd: [{
                type: HostBinding,
                args: ["class.cds--text-input--md"]
            }], isSizelg: [{
                type: HostBinding,
                args: ["class.cds--text-input--lg"]
            }], sizeSm: [{
                type: HostBinding,
                args: ["class.cds--layout--size-sm"]
            }], sizeMd: [{
                type: HostBinding,
                args: ["class.cds--layout--size-md"]
            }], sizelg: [{
                type: HostBinding,
                args: ["class.cds--layout--size-lg"]
            }], invalid: [{
                type: HostBinding,
                args: ["class.cds--text-input--invalid"]
            }, {
                type: Input
            }], warn: [{
                type: HostBinding,
                args: ["class.cds--text-input--warning"]
            }, {
                type: Input
            }], skeleton: [{
                type: HostBinding,
                args: ["class.cds--skeleton"]
            }, {
                type: Input
            }], isLightTheme: [{
                type: HostBinding,
                args: ["class.cds--text-input--light"]
            }], getInvalidAttribute: [{
                type: HostBinding,
                args: ["attr.data-invalid"]
            }] } });
