// carbon-components-angular/esm2020/notification/base-notification.component.mjs
import { Output, EventEmitter, HostBinding, Component } from "@angular/core";
import { isObservable } from "rxjs";
import * as i0 from "@angular/core";
import * as i1 from "./notification-display.service";
import * as i2 from "carbon-components-angular/i18n";
/**
 * Base class for `Notification`, `ActionableNotification`, & `Toast`
 * consisting of common functionality
 *
 * Get started with importing the module:
 *
 * ```typescript
 * import { NotificationModule } from 'carbon-components-angular';
 * ```
 *
 * [See demo](../../?path=/story/components-notification--inline-notification)
 */
export class BaseNotification {
    constructor(notificationDisplayService, i18n) {
        this.notificationDisplayService = notificationDisplayService;
        this.i18n = i18n;
        /**
         * Emits on close.
         */
        this.close = new EventEmitter();
        // Get icon name(value) for service based on the notification type (key)
        this.iconDictionary = {
            "error": "error--filled",
            "info": "information--filled",
            "info-square": "information--square--filled",
            "success": "checkmark--filled",
            "warning": "warning--filled",
            "warning-alt": "warning--alt--filled"
        };
        this.defaultNotificationObj = {
            title: "",
            message: "",
            type: "info",
            showClose: true,
            closeLabel: this.i18n.get("NOTIFICATION.CLOSE_BUTTON"),
            role: "status"
        };
        this._notificationObj = Object.assign({}, this.defaultNotificationObj);
    }
    /**
     * Set role attribute for component
     * `Status` is default for inline-notification & toast component
     * `alertdialog` is default for actionable-notification
     */
    get roleAttr() {
        return this._notificationObj.role;
    }
    /**
     * Emits close event.
     */
    onClose() {
        this.close.emit();
    }
    onClick(action, event) {
        if (!action.click) {
            return;
        }
        if (isObservable(action.click)) {
            action.click.next({ event, action });
        }
        else {
            action.click({ event, action });
        }
    }
    destroy() {
        this.notificationDisplayService.close(this);
    }
}
BaseNotification.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: BaseNotification, deps: [{ token: i1.NotificationDisplayService }, { token: i2.I18n }], target: i0.ɵɵFactoryTarget.Component });
BaseNotification.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.3.0", type: BaseNotification, selector: "ng-component", outputs: { close: "close" }, host: { properties: { "attr.role": "this.roleAttr" } }, ngImport: i0, template: "", isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: BaseNotification, decorators: [{
            type: Component,
            args: [{
                    template: ""
                }]
        }], ctorParameters: function () { return [{ type: i1.NotificationDisplayService }, { type: i2.I18n }]; }, propDecorators: { roleAttr: [{
                type: HostBinding,
                args: ["attr.role"]
            }], close: [{
                type: Output
            }] } });
