// carbon-components-angular/esm2020/dialog/dialog.module.mjs
// modules
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
// imports
import { DialogService } from "./dialog.service";
import { Dialog } from "./dialog.component";
import { DialogDirective } from "./dialog.directive";
import { OverflowMenu } from "./overflow-menu/overflow-menu.component";
import { OverflowMenuPane } from "./overflow-menu/overflow-menu-pane.component";
import { OverflowMenuCustomPane } from "./overflow-menu/overflow-menu-custom-pane.component";
import { OverflowMenuDirective } from "./overflow-menu/overflow-menu.directive";
import { OverflowMenuOption } from "./overflow-menu/overflow-menu-option.component";
import { I18nModule } from "carbon-components-angular/i18n";
import { PlaceholderModule } from "carbon-components-angular/placeholder";
import { ExperimentalModule } from "carbon-components-angular/experimental";
import { UtilsModule } from "carbon-components-angular/utils";
import { IconModule } from "carbon-components-angular/icon";
import { ButtonModule } from "carbon-components-angular/button";
import { TooltipModule } from "carbon-components-angular/tooltip";
import * as i0 from "@angular/core";
export class DialogModule {
}
DialogModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DialogModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.3.0", ngImport: i0, type: DialogModule, declarations: [Dialog,
        OverflowMenu,
        OverflowMenuPane,
        OverflowMenuCustomPane,
        DialogDirective,
        OverflowMenuDirective,
        OverflowMenuOption], imports: [CommonModule,
        I18nModule,
        PlaceholderModule,
        ExperimentalModule,
        UtilsModule,
        IconModule,
        ButtonModule,
        TooltipModule], exports: [Dialog,
        OverflowMenu,
        OverflowMenuPane,
        OverflowMenuCustomPane,
        DialogDirective,
        OverflowMenuDirective,
        OverflowMenuOption] });
DialogModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DialogModule, providers: [DialogService], imports: [CommonModule,
        I18nModule,
        PlaceholderModule,
        ExperimentalModule,
        UtilsModule,
        IconModule,
        ButtonModule,
        TooltipModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.3.0", ngImport: i0, type: DialogModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        Dialog,
                        OverflowMenu,
                        OverflowMenuPane,
                        OverflowMenuCustomPane,
                        DialogDirective,
                        OverflowMenuDirective,
                        OverflowMenuOption
                    ],
                    exports: [
                        Dialog,
                        OverflowMenu,
                        OverflowMenuPane,
                        OverflowMenuCustomPane,
                        DialogDirective,
                        OverflowMenuDirective,
                        OverflowMenuOption
                    ],
                    providers: [DialogService],
                    imports: [
                        CommonModule,
                        I18nModule,
                        PlaceholderModule,
                        ExperimentalModule,
                        UtilsModule,
                        IconModule,
                        ButtonModule,
                        TooltipModule
                    ]
                }]
        }] });
